megatron/core/models/deepseekv2/transformer_layer.py

```diff
@@ -0,0 +1,124 @@
+from abc import ABC
+from dataclasses import dataclass, field
+from typing import Dict, Optional, Union
+
+import torch
+
+from megatron.core import parallel_state
+from megatron.core.dist_checkpointing.mapping import ShardedStateDict
+from megatron.core.dist_checkpointing.utils import apply_prefix_mapping
+from megatron.core.transformer.enums import AttnMaskType
+from megatron.core.transformer.transformer_layer import TransformerLayer, TransformerLayerSubmodules
+from megatron.core.transformer.identity_op import IdentityFuncOp, IdentityOp
+from megatron.core.transformer.module import MegatronModule
+from megatron.core.transformer.spec_utils import ModuleSpec, build_module
+from megatron.core.transformer.transformer_config import TransformerConfig
+from megatron.core.utils import make_viewless_tensor
+
+
+class DeekSeekv2TransformerLayer(TransformerLayer):
+    """A single transformer layer for Deepseekv2.
+
+    Transformer layer takes input with size [s, b, h] and returns an
+    output of the same size.
+    """
+    def __init__(
+        self,
+        config: TransformerConfig,
+        submodules: TransformerLayerSubmodules,
+        layer_number: int = 1,
+        hidden_dropout: float = None,
+    ):
+        super().__init__(config=config,
+                         submodules=submodules,
+                         layer_number=layer_number,
+                         hidden_dropout=hidden_dropout)
+        self.world_size = parallel_state.get_tensor_model_parallel_world_size()
+    
+    def forward(
+        self,
+        hidden_states,
+        position_ids,
+        attention_mask,
+        context=None,
+        context_mask=None,
+        rotary_pos_emb=None,
+        inference_params=None,
+        packed_seq_params=None,
+    ):
+        # hidden_states: [s, b, h]
+
+        # Residual connection.
+        residual = hidden_states
+
+        # Optional Input Layer norm
+        input_layernorm_output = self.input_layernorm(hidden_states)
+
+        # Self attention.
+        attention_output_with_bias = self.self_attention(
+            input_layernorm_output,
+            position_ids=position_ids,
```

Comment:
It seems the main difference between the `forward` functions of `DeekSeekv2TransformerLayer` and `TransformerLayer` is the extra argument of `position_ids` in this function call. I am wondering if we can refactor it to reduce the code duplication.